### PR TITLE
Updated BackoffSupervisor to Akka JVM 2.5.18

### PR DESCRIPTION
**Pull request by ismaelhamed**
Updated BackoffSupervisor to Akka JVM 2.5.18

* BackoffSupervisor notifies sender when child stop
* Fixed BackoffSupervisor.CalculateDelay (and try to settle #3318)
* Added maximum restart attempts to BackoffSupervisor

### CODE REVIEW

File: src/core/Akka.Tests/Pattern/BackoffSupervisorSpec.cs [hunk 1]
```diff
@@ -258,5 +260,192 @@ public void BackoffSupervisor_must_support_manual_reset()
...
+            public IEnumerator<object[]> GetEnumerator() => delayTable.GetEnumerator();
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        [Fact]
+        public void BackoffSupervisor_stop_restarting_the_child_after_reaching_maxNrOfRetries_limit_using_BackOff_OnStop()
+        {
+            var supervisor = Create(OnStopOptions(maxNrOfRetries: 2));
+
+            IActorRef WaitForChild(int n = 0)
```
Comment (ismaelhamed) [inline src/core/Akka.Tests/Pattern/BackoffSupervisorSpec.cs]: Do we have anything like `scalatest.concurrent.Eventually`? That would simplify this piece of code.

Comment (Aaronontheweb) [reply src/core/Akka.Tests/Pattern/BackoffSupervisorSpec.cs]: `AwaitCondition` ?

**Comment by Aaronontheweb** [comment]
@ismaelhamed is this still a work in progress? or is it ready for review?

File: src/core/Akka.Tests/Pattern/BackoffSupervisorSpec.cs (hunk 1 above)
Comment (ismaelhamed) [reply src/core/Akka.Tests/Pattern/BackoffSupervisorSpec.cs]: Exactly that.

**Comment by ismaelhamed** [comment]
It's ready for review. I've just updated the tests with the `AwaitCondition`.

**Comment by ismaelhamed** [comment]
@Aaronontheweb I have another PR ready, that builds on top of this one, with [#25933](https://github.com/akka/akka/pull/25933). This should put the BackoffSupervisor on a par with the JVM.

File: src/core/Akka.Tests/Pattern/BackoffSupervisorSpec.cs [hunk 2]
```diff
@@ -258,5 +260,184 @@ public void BackoffSupervisor_must_support_manual_reset()
...
+                ExpectTerminated(c1);
+
+                AwaitAssert(() =>
+                {
+                    supervisor.Tell(BackoffSupervisor.GetRestartCount.Instance);
+                    ExpectMsg<BackoffSupervisor.RestartCount>().Count.Should().Be(1);
+                });
+
+                supervisor.Tell("boom"); //this will be sent to deadLetters
+                ExpectNoMsg(500);
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Pattern/BackoffSupervisorSpec.cs]: LGTM

File: src/core/Akka/Pattern/BackoffOptions.cs [hunk 3]
```diff
@@ -23,9 +23,25 @@ public static class Backoff
         /// <param name="minBackoff">Minimum (initial) duration until the child actor will started again, if it is terminated</param>
         /// <param name="maxBackoff">The exponential back-off is capped to this duration</param>
         /// <param name="randomFactor">After calculation of the exponential back-off an additional random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay. In order to skip this additional delay pass in `0`.</param>
+        [Obsolete("Use the overloaded one which accepts maxNrOfRetries instead.")]
```
Comment (Aaronontheweb) [inline src/core/Akka/Pattern/BackoffOptions.cs]: LGTM

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt [hunk 4]
```diff
@@ -3754,22 +3755,29 @@ namespace Akka.Pattern
...
         public abstract Akka.Pattern.BackoffOptions WithManualReset();
+        public abstract Akka.Pattern.BackoffOptions WithMaxNrOfRetries(int maxNrOfRetries);
+        public abstract Akka.Pattern.BackoffOptions WithReplyWhileStopped(object replyWhileStopped);
         public abstract Akka.Pattern.BackoffOptions WithSupervisorStrategy(Akka.Actor.OneForOneStrategy supervisorStrategy);
     }
     public sealed class BackoffSupervisor : Akka.Pattern.BackoffSupervisorBase
     {
         public BackoffSupervisor(Akka.Actor.Props childProps, string childName, System.TimeSpan minBackoff, System.TimeSpan maxBackoff, double randomFactor) { }
-        public BackoffSupervisor(Akka.Actor.Props childProps, string childName, System.TimeSpan minBackoff, System.TimeSpan maxBackoff, Akka.Pattern.IBackoffReset reset, double randomFactor, Akka.Actor.SupervisorStrategy strategy) { }
+        public BackoffSupervisor(Akka.Actor.Props childProps, string childName, System.TimeSpan minBackoff, System.TimeSpan maxBackoff, Akka.Pattern.IBackoffReset reset, double randomFactor, Akka.Actor.SupervisorStrategy strategy, object replyWhileStopped = null) { }
```
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt]: So this looked like a public breaking API change to me, but upon further review - this is actually calling an `internal` constructor belonging to the base class and can't be directly exposed to end users.

**Comment by Aaronontheweb** [comment]
@ismaelhamed done!

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt (hunk 4 above)
Comment (ismaelhamed) [reply src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt]: @Aaronontheweb So it is considered a breaking change even when just adding optional parameters to a method signature?

**Comment by Aaronontheweb** [comment]
Correct

**Pull request merged**